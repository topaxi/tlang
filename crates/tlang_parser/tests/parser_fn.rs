use indoc::indoc;

mod common;

#[test]
fn test_function_declaration() {
    assert_parser_snapshot!("fn foo() { bar(); 1 + 2; }");
}

#[test]
fn test_function_declaration_with_parameters() {
    assert_parses!("fn foo(x, y) { 1 + 2; }");
}

#[test]
fn test_nameless_function_expressions() {
    assert_parses!("let x = fn() { 1 + 2; };");
    assert_parses!("let x = fn(x, y) { 1 + 2; };");
}

#[test]
fn test_function_expression_without_name_no_argument_parenthesis() {
    assert_parses!("let x = fn { 1 + 2; };");
}

#[test]
fn test_function_expressions() {
    assert_parses!("let x = fn foo() { 1 + 2; };");
    assert_parses!("let x = fn foo(x, y) { 1 + 2; };");
}

#[test]
fn test_explicit_return_statements() {
    assert_parses!("fn foo() { return 1 + 2; }");
    assert_parses!("fn foo() { return; }");
    assert_parses!("let x = fn() { return 1 + 2; };");
}

#[test]
fn test_implicit_return_expressions() {
    assert_parses!("fn foo() { 1 + 2 }");
    assert_parses!("let x = fn() { 1 + 2 };");
    assert_parser_snapshot!("fn foo() { let x = 1; x }");
}

#[test]
fn test_recursive_factorial_functional_definition() {
    assert_parser_snapshot!(indoc! {"
        fn factorial(0) { return 1; }
        fn factorial(n) { return n * factorial(n - 1); }
    "});
}

#[test]
fn test_recursive_sum() {
    assert_parser_snapshot!(indoc! {"
        fn sum([]) { 0 }
        fn sum([x, ...xs]) { x + sum(xs) }
    "});
}

#[test]
fn test_functional_function_declaration_with_comments_inbetween() {
    assert_parser_snapshot!(indoc! {"
        fn foo(1) { 1 }
        // comment
        fn foo(n) { n * 2 }
    "});
}

#[test]
fn test_explicit_tail_recursive_call() {
    assert_parser_snapshot!(indoc! {"
        fn factorial(n) { factorial(n, 1) }
        fn factorial(0, acc) { acc }
        fn factorial(n, acc) { rec factorial(n - 1, n * acc) }
    "});
}

#[test]
fn test_foldl_impl() {
    assert_parser_snapshot!(indoc! {"
        fn foldl([], acc, _) { acc }
        fn foldl([x, ...xs], acc, f) { rec foldl(xs, f(acc, x), f) }
    "});
}

#[test]
fn test_function_declarations_with_guard() {
    assert_parser_snapshot!(indoc! {"
        fn filter([], f) { [] }
        fn filter([x, ...xs], f) if f(x) { [x, ...filter(xs, f)] }
        fn filter([x, ...xs], f) { rec filter(xs, f) }
    "});
}

#[test]
fn test_function_declarations_with_let_guard() {
    assert_parser_snapshot!(indoc! {"
        fn filter_map([], f) { [] }
        fn filter_map([x, ...xs], f) if let y = f(x) { [y, ...filter_map(xs, f)] }
        fn filter_map([x, ...xs], f) { rec filter_map(xs, f) }
    "});
}

#[test]
#[should_panic]
fn test_parameter_list_mandatory_comma() {
    parse!("fn foo(x y) {}");
}

#[test]
fn test_list_matching_wildcard() {
    assert_parser_snapshot!("fn tail([_, ...xs]) { xs }");
}

#[test]
fn test_fn_expression_in_function_completion_position() {
    assert_parser_snapshot!("fn foo() { fn bar() {} }");
}
