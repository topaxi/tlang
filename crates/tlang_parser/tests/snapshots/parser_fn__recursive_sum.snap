---
source: crates/tlang_parser/tests/parser_fn.rs
expression: "parse!(indoc! {\n        \"\n        fn sum([]) { 0 }\n        fn sum([x, ...xs]) { x + sum(xs) }\n    \"\n    })"
---
Node(
  ast_node: Legacy(Module([
    Stmt(
      kind: FunctionDeclarations([
        Stmt(
          kind: FunctionDeclaration(FunctionDeclaration(
            id: SymbolId(1),
            name: Expr(
              kind: Identifier(Ident(
                name: "sum",
                span: Span(
                  start: LineColumn(
                    line: 0,
                    column: 3,
                  ),
                  end: LineColumn(
                    line: 0,
                    column: 6,
                  ),
                ),
              )),
              symbol_table: None,
              span: Span(
                start: LineColumn(
                  line: 0,
                  column: 3,
                ),
                end: LineColumn(
                  line: 0,
                  column: 6,
                ),
              ),
            ),
            parameters: [
              FunctionParameter(
                pattern: Pattern(
                  kind: List([]),
                  span: Span(
                    start: LineColumn(
                      line: 0,
                      column: 7,
                    ),
                    end: LineColumn(
                      line: 0,
                      column: 9,
                    ),
                  ),
                ),
                type_annotation: None,
                span: Span(
                  start: LineColumn(
                    line: 0,
                    column: 7,
                  ),
                  end: LineColumn(
                    line: 0,
                    column: 9,
                  ),
                ),
              ),
            ],
            guard: None,
            return_type_annotation: None,
            body: Expr(
              kind: Block([], Some(Expr(
                kind: Literal(Integer(0)),
                symbol_table: None,
                span: Span(
                  start: LineColumn(
                    line: 0,
                    column: 13,
                  ),
                  end: LineColumn(
                    line: 0,
                    column: 14,
                  ),
                ),
              ))),
              symbol_table: None,
              span: Span(
                start: LineColumn(
                  line: 0,
                  column: 0,
                ),
                end: LineColumn(
                  line: 0,
                  column: 0,
                ),
              ),
            ),
          )),
          span: Span(
            start: LineColumn(
              line: 0,
              column: 0,
            ),
            end: LineColumn(
              line: 0,
              column: 0,
            ),
          ),
          symbol_table: None,
        ),
        Stmt(
          kind: FunctionDeclaration(FunctionDeclaration(
            id: SymbolId(1),
            name: Expr(
              kind: Identifier(Ident(
                name: "sum",
                span: Span(
                  start: LineColumn(
                    line: 1,
                    column: 4,
                  ),
                  end: LineColumn(
                    line: 1,
                    column: 7,
                  ),
                ),
              )),
              symbol_table: None,
              span: Span(
                start: LineColumn(
                  line: 1,
                  column: 4,
                ),
                end: LineColumn(
                  line: 1,
                  column: 7,
                ),
              ),
            ),
            parameters: [
              FunctionParameter(
                pattern: Pattern(
                  kind: List([
                    Pattern(
                      kind: Identifier(
                        id: SymbolId(2),
                        name: Ident(
                          name: "x",
                          span: Span(
                            start: LineColumn(
                              line: 1,
                              column: 9,
                            ),
                            end: LineColumn(
                              line: 1,
                              column: 10,
                            ),
                          ),
                        ),
                      ),
                      span: Span(
                        start: LineColumn(
                          line: 1,
                          column: 9,
                        ),
                        end: LineColumn(
                          line: 1,
                          column: 10,
                        ),
                      ),
                    ),
                    Pattern(
                      kind: Rest(Pattern(
                        kind: Identifier(
                          id: SymbolId(3),
                          name: Ident(
                            name: "xs",
                            span: Span(
                              start: LineColumn(
                                line: 1,
                                column: 15,
                              ),
                              end: LineColumn(
                                line: 1,
                                column: 17,
                              ),
                            ),
                          ),
                        ),
                        span: Span(
                          start: LineColumn(
                            line: 1,
                            column: 15,
                          ),
                          end: LineColumn(
                            line: 1,
                            column: 17,
                          ),
                        ),
                      )),
                      span: Span(
                        start: LineColumn(
                          line: 0,
                          column: 0,
                        ),
                        end: LineColumn(
                          line: 0,
                          column: 0,
                        ),
                      ),
                    ),
                  ]),
                  span: Span(
                    start: LineColumn(
                      line: 1,
                      column: 8,
                    ),
                    end: LineColumn(
                      line: 1,
                      column: 18,
                    ),
                  ),
                ),
                type_annotation: None,
                span: Span(
                  start: LineColumn(
                    line: 1,
                    column: 8,
                  ),
                  end: LineColumn(
                    line: 1,
                    column: 18,
                  ),
                ),
              ),
            ],
            guard: None,
            return_type_annotation: None,
            body: Expr(
              kind: Block([], Some(Expr(
                kind: BinaryOp(
                  op: Add,
                  lhs: Expr(
                    kind: Identifier(Ident(
                      name: "x",
                      span: Span(
                        start: LineColumn(
                          line: 1,
                          column: 22,
                        ),
                        end: LineColumn(
                          line: 1,
                          column: 23,
                        ),
                      ),
                    )),
                    symbol_table: None,
                    span: Span(
                      start: LineColumn(
                        line: 1,
                        column: 22,
                      ),
                      end: LineColumn(
                        line: 1,
                        column: 23,
                      ),
                    ),
                  ),
                  rhs: Expr(
                    kind: Call(
                      function: Expr(
                        kind: Identifier(Ident(
                          name: "sum",
                          span: Span(
                            start: LineColumn(
                              line: 1,
                              column: 26,
                            ),
                            end: LineColumn(
                              line: 1,
                              column: 29,
                            ),
                          ),
                        )),
                        symbol_table: None,
                        span: Span(
                          start: LineColumn(
                            line: 0,
                            column: 0,
                          ),
                          end: LineColumn(
                            line: 0,
                            column: 0,
                          ),
                        ),
                      ),
                      arguments: [
                        Expr(
                          kind: Identifier(Ident(
                            name: "xs",
                            span: Span(
                              start: LineColumn(
                                line: 1,
                                column: 30,
                              ),
                              end: LineColumn(
                                line: 1,
                                column: 32,
                              ),
                            ),
                          )),
                          symbol_table: None,
                          span: Span(
                            start: LineColumn(
                              line: 1,
                              column: 30,
                            ),
                            end: LineColumn(
                              line: 1,
                              column: 32,
                            ),
                          ),
                        ),
                      ],
                    ),
                    symbol_table: None,
                    span: Span(
                      start: LineColumn(
                        line: 1,
                        column: 26,
                      ),
                      end: LineColumn(
                        line: 1,
                        column: 33,
                      ),
                    ),
                  ),
                ),
                symbol_table: None,
                span: Span(
                  start: LineColumn(
                    line: 1,
                    column: 22,
                  ),
                  end: LineColumn(
                    line: 1,
                    column: 33,
                  ),
                ),
              ))),
              symbol_table: None,
              span: Span(
                start: LineColumn(
                  line: 0,
                  column: 0,
                ),
                end: LineColumn(
                  line: 0,
                  column: 0,
                ),
              ),
            ),
          )),
          span: Span(
            start: LineColumn(
              line: 0,
              column: 0,
            ),
            end: LineColumn(
              line: 0,
              column: 0,
            ),
          ),
          symbol_table: None,
        ),
      ]),
      span: Span(
        start: LineColumn(
          line: 0,
          column: 0,
        ),
        end: LineColumn(
          line: 1,
          column: 35,
        ),
      ),
      symbol_table: None,
    ),
  ])),
  symbol_table: None,
  span: Span(
    start: LineColumn(
      line: 0,
      column: 0,
    ),
    end: LineColumn(
      line: 0,
      column: 0,
    ),
  ),
)
